lightning-invoice/src/payment.rs

```diff
@@ -2424,4 +2424,143 @@ mod tests {
+		let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
+		let nodes = create_network(3, &node_cfgs, &node_chanmgrs);
+
+		let chan_1_scid = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 10_000_000, 0, channelmanager::provided_init_features(), channelmanager::provided_init_features()).0.contents.short_channel_id;
+		let chan_2_scid = create_announced_chan_between_nodes_with_value(&nodes, 1, 2, 10_000_000, 0, channelmanager::provided_init_features(), channelmanager::provided_init_features()).0.contents.short_channel_id;
+
+		let mut route = Route {
+			paths: vec![
+				vec![RouteHop {
+					pubkey: nodes[1].node.get_our_node_id(),
+					node_features: channelmanager::provided_node_features(),
+					short_channel_id: chan_1_scid,
+					channel_features: channelmanager::provided_channel_features(),
+					fee_msat: 0,
+					cltv_expiry_delta: 100,
+				}, RouteHop {
+					pubkey: nodes[2].node.get_our_node_id(),
+					node_features: channelmanager::provided_node_features(),
+					short_channel_id: chan_2_scid,
+					channel_features: channelmanager::provided_channel_features(),
+					fee_msat: 100_000_000,
+					cltv_expiry_delta: 100,
+				}],
+				vec![RouteHop {
+					pubkey: nodes[1].node.get_our_node_id(),
+					node_features: channelmanager::provided_node_features(),
+					short_channel_id: chan_1_scid,
+					channel_features: channelmanager::provided_channel_features(),
+					fee_msat: 0,
+					cltv_expiry_delta: 100,
+				}, RouteHop {
+					pubkey: nodes[2].node.get_our_node_id(),
+					node_features: channelmanager::provided_node_features(),
+					short_channel_id: chan_2_scid,
+					channel_features: channelmanager::provided_channel_features(),
+					fee_msat: 100_000_000,
+					cltv_expiry_delta: 100,
+				}]
+			],
+			payment_params: Some(PaymentParameters::from_node_id(nodes[2].node.get_our_node_id())),
+		};
+		let router = ManualRouter(RefCell::new(VecDeque::new()));
+		router.expect_find_route(Ok(route.clone()));
+		// On retry, we'll only be asked for one path
+		route.paths.remove(1);
+		router.expect_find_route(Ok(route.clone()));
+
+		let expected_events: RefCell<VecDeque<&dyn Fn(&Event)>> = RefCell::new(VecDeque::new());
+		let event_handler = |event: &Event| {
+			let event_checker = expected_events.borrow_mut().pop_front().unwrap();
+			event_checker(event);
+		};
+		let invoice_payer = InvoicePayer::new(nodes[0].node, router, nodes[0].logger, event_handler, Retry::Attempts(1));
+
+		assert!(invoice_payer.pay_invoice(&create_invoice_from_channelmanager_and_duration_since_epoch(
+			&nodes[1].node, nodes[1].keys_manager, Currency::Bitcoin, Some(100_010_000), "Invoice".to_string(),
+			duration_since_epoch(), 3600).unwrap())
+			.is_ok());
+		let entries=nodes.iter().enumerate().map(|(i, n)| (n.node.node_id(), i));
+		let node_map:HashMap<PublicKey, usize>=entries.collect();
+		handle_message_send_events(&nodes[0], &node_map, &nodes);
```

Comment:
As mentioned by @TheBlueMatt in chat, it would probably make sense to reuse the `commitment_signed_dance!` macro here, instead of introducing a new recursive helper function.

https://github.com/lightningdevkit/rust-lightning/blob/7bc52aa62cb3c32f51b242ad3fa0ae16aee80cec/lightning/src/ln/functional_test_utils.rs#L1146-L1228

You could consider using this in combination with a loop here, but likely it would make it harder to follow who sends to whom. Since the macro allows to remove the intermediary steps between HTLC updates/failures, this should increase readability quite a bit already.